src/app/clusters/camera-av-stream-management-server/camera-av-stream-management-server.h

```diff
@@ -0,0 +1,656 @@
+     *   produced; otherwise, the command SHALL be rejected with an appropriate
+     *   error.
+     */
+    virtual Protocols::InteractionModel::Status SnapshotStreamAllocate(ImageCodecEnum imageCodec, const uint16_t frameRate,
+                                                                       const uint32_t bitRate,
+                                                                       const VideoResolutionStruct & minResolution,
+                                                                       const VideoResolutionStruct & maxResolution,
+                                                                       const uint8_t quality, uint16_t & outStreamID) = 0;
+
+    /**
+     *   @brief Handle Command Delegate for Snapshot stream deallocation.
+     *
+     *   @param streamID       Indicates the streamID to deallocate.
+     *
+     *   @return Success if the stream deallocation is successful; otherwise, the command SHALL be rejected with an appropriate
+     *   error.
+     */
+    virtual Protocols::InteractionModel::Status SnapshotStreamDeallocate(const uint16_t streamID) = 0;
+
+    /**
+     *   @brief Handle Command Delegate for SetStreamPriorities.
+     *
+     *   @param streamPriorities    Indicates the list of stream priorities to set, in decreasing order of priorities.
+     *
+     *   @return Success if the setting of stream priorities is successful; otherwise, the command SHALL be rejected with an
+     * appropriate error.
+     */
+    virtual Protocols::InteractionModel::Status
+    SetStreamPriorities(const DataModel::DecodableList<StreamUsageEnum> streamPriorities) = 0;
+
+    /**
+     *   @brief Handle Command Delegate for CaptureSnapshot.
+     *
+     *   @param streamID          Indicates the streamID representing the shapshot stream.
+     *
+     *   @param videoResolution   Indicates the preferred resolution of the snapshot image.
+     *
+     *   @return Success if the processing of the Command is successful; otherwise, the command SHALL be rejected with an
+     * appropriate error.
+     */
+    virtual Protocols::InteractionModel::Status CaptureSnapshot(const uint16_t streamID,
+                                                                const VideoResolutionStruct & videoResolution) = 0;
+
+    /**
+     *  Delegate functions to load the allocated video, audio, and snapshot streams.
+     *  The delegate application is responsible for creating and persisting
+     *  these streams(based on the Allocation commands). These Load APIs would be
+     *  used to load the pre-allocated stream context information into the cluster server list,
+     *  at initialization.
+     *  Once loaded, the cluster server would be serving Reads on these
+     *  attributes. The list is updatable via the Add/Remove functions for the
+     *  respective streams.
+     */
+    virtual Protocols::InteractionModel::Status
+    LoadAllocatedVideoStreams(std::vector<VideoStreamStruct> & allocatedVideoStreams) = 0;
```

Comment:
Has there been a conscious decision to use `std::vector` in this code? Most core code avoids using it.

Reply:
I have used vectors here mainly for ease of use for managing dynamic lists, and especially, since cameras are expected to be relatively resourceful. What is the preferred approach to represent dynamic lists that are owned?

Reply:
If the objects are very large/non-copyable, then we might use std::vector<std::unique_ptr<VideoStreamStruct>>?
 
std::unique_ptr clearly expresses unique ownership: the vector “owns” the objects. This helps avoid memory leaks and clarifies that nothing else should be responsible for those objects’ lifetimes.

Reply:
Allocation should be infrequent and the number of streams should be relatively small. Keeping it as a vector of structs seems simpler to manage.